File: demo/KSYLiveDemo/KSYDemoUI/KSYCtrlView.m
//
//  controlView.m
//  KSYDemo
//
//  Created by 孙健 on 16/4/6.
//  Copyright © 2016年 孙健. All rights reserved.
//

#import "KSYCtrlView.h"
#import "KSYStateLableView.h"

@interface KSYCtrlView () {
    KSYUIView * _curSubMenuView;
}

@end

@implementation KSYCtrlView

- (id) initWithMenu:(NSArray *) menuNames {
    self = [super init];
    _btnFlash  =  [self addButton:@"闪光灯" ];
    _btnCameraToggle =  [self addButton:@"前后摄像头" ];
    _btnQuit   =  [self addButton:@"退出" ];
    _lblNetwork=  [self addLable:@""  ];
    _btnStream =  [self addButton:@"推流"  ];
    _btnCapture=  [self addButton:@"采集"  ];
    _lblStat   =  [[KSYStateLableView alloc] init];
    [self addSubview:_lblStat];
    // format
    _lblNetwork.textAlignment = NSTextAlignmentCenter;
    
    // add menu
    NSMutableArray * btnArray = [[NSMutableArray alloc] init];
    for (NSString * name in menuNames){
        [btnArray addObject: [self addButton:name] ];
    }
    _menuBtns = [NSArray arrayWithArray:btnArray];
    
    _backBtn   = [self addButton:@"菜单"
                          action:@selector(onBack:)];
    _backBtn.hidden = YES;
    _curSubMenuView = nil;
    return self;
}

- (void) layoutUI {
    [super layoutUI];
    if ( self.width <self.height ){
        self.yPos =self.gap*5; // skip status bar
    }
    [self putRow: @[_backBtn, _btnFlash, _btnCameraToggle,_btnQuit] ];
    
    [self putRow: _menuBtns ];
    [self hideMenuBtn:!_backBtn.hidden];
    
    self.yPos -= self.btnH;
    CGFloat freeHgt = self.height - self.yPos - self.btnH - self.gap;
    _lblStat.frame = CGRectMake( self.gap, self.yPos, self.winWdt - self.gap*2, freeHgt);
    self.yPos += freeHgt;
    
    // put at bottom
    [self putRow3:_btnCapture
              and:_lblNetwork
              and:_btnStream];
    if (_curSubMenuView) {
        _curSubMenuView.frame = _lblStat.frame;
        [_curSubMenuView layoutUI];
    }
}

- (void)hideMenuBtn: (BOOL) bHide {
    _backBtn.hidden   = !bHide; // 返回
    // hide menu
    for (UIButton * btn in _menuBtns){
        btn.hidden = bHide;
    }
}

- (IBAction)onBack:(id)sender {
    if (_curSubMenuView){
        _curSubMenuView.hidden = YES;
    }
    [self hideMenuBtn:NO];
}
- (void) showSubMenuView: (KSYUIView*) view {
    _curSubMenuView = view;
    [self hideMenuBtn:YES];
    view.hidden = NO;
    view.frame = _lblStat.frame;
    [view layoutUI];
}
@end

File: demo/KSYLiveDemo_Swift/KSYDemoUI_Swift/subviews/KSYCtrlView.swift
//
//  KSYCtrlView.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 17/1/10.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

class KSYCtrlView: KSYUIView {

    var btnFlash: UIButton?
    var btnCameraToggle: UIButton?
    var btnQuit: UIButton?
    var btnStream: UIButton?
    var btnCapture: UIButton?

    var lblStat: KSYStateLableView?
    var lblNetwork: UILabel?
    
    //背景音乐
    //图像和美颜相关
    //声音相关: 混音 / 混响 / 耳返等
    //其他功能: 比如截屏
    var menuBtns: [UIButton]?
    //返回菜单页面
    var backBtn: UIButton?
    
    private
    var _curSubMenuView: KSYUIView?
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    init(menuNames: [String]) {
        super.init()
        btnFlash = addButton(title: "闪光灯")
        btnCameraToggle = addButton(title: "前后摄像头")
        btnQuit = addButton(title: "退出")
        lblNetwork = addLabel(title: "")
        btnStream = addButton(title: "推流")
        btnCapture = addButton(title: "采集")
        lblStat = KSYStateLableView()
        addSubview(lblStat!)
        
        // format
        lblNetwork?.textAlignment = .center
        
        // add menu
        var btnArray = Array<UIButton>()
        for name: String in menuNames {
            btnArray.append(addButton(title: name))
        }
        
        menuBtns = btnArray
        
        backBtn = addButton(title: "菜单", action: #selector(onBack(sender:)))
        backBtn?.isHidden = true
        _curSubMenuView = nil
    }
    
    override func layoutUI() {
        super.layoutUI()
        if width < height {
            yPos = gap * 5  // skip status
        }
        
        putRow(subV: [btnQuit!, btnFlash!, btnCameraToggle!, backBtn!])
        
        putRow(subV: menuBtns!)
        hideMenuBtn(bHide: !backBtn!.isHidden)
        
        yPos -= btnH
        let freeHgt = height - yPos - btnH - gap
        lblStat?.frame = CGRect.init(x: gap, y: yPos, width: winWdt - gap * 2, height: freeHgt)
        yPos += freeHgt
        
        // put at bottom
        putRow3(subV0: btnCapture!, and: lblNetwork!, and: btnStream!)
        
        if let _ = _curSubMenuView {
            _curSubMenuView!.frame = lblStat!.frame
            _curSubMenuView!.layoutUI()
        }
    }
    
    func hideMenuBtn(bHide: Bool) {
        backBtn!.isHidden = !bHide // 返回
        // hide menu
        for btn in menuBtns! {
            btn.isHidden = bHide
        }
    }
    
    func onBack(sender: UIButton) {
        if let _ = _curSubMenuView {
            _curSubMenuView?.isHidden = true
        }
        hideMenuBtn(bHide: false)
    }
    
    func showSubMenuView(view: KSYUIView) {
        _curSubMenuView = view
        hideMenuBtn(bHide: true)
        view.isHidden = false
        view.frame = lblStat!.frame
        view.layoutUI()
    }
}
